packages/components/src/components/select/select.component.ts

```diff
@@ -155,31 +169,91 @@ class Select
   /** @internal */
   private initialSelectedOption: Option | null = null;
 
-  private getAllValidOptions(): Array<Option> {
-    return Array.from(this.slottedListboxes[0]?.querySelectorAll(`${OPTION_TAG_NAME}:not([disabled])`) || []);
+  constructor() {
+    super();
+
+    this.addEventListener(LIFE_CYCLE_EVENTS.CREATED, this.handleCreatedEvent);
+    this.addEventListener(LIFE_CYCLE_EVENTS.DESTROYED, this.handleDestroyEvent);
+    // This must be initialized after the destroyed event listener
+    // to keep the element in the itemStore in order to move the focus correctly
+    this.itemsStore = new ElementStore<Option>(this, {
+      isValidItem: this.isValidItem,
+    });
   }
 
-  private getFirstValidOption(): Option | null {
-    return this.slottedListboxes[0]?.querySelector(`${OPTION_TAG_NAME}:not([disabled])`);
+  override connectedCallback(): void {
+    super.connectedCallback();
+
+    this.loop = 'false';
+    this.initialFocus = 0;
   }
 
-  private getLastValidOption(): Option | null {
-    const options = this.getAllValidOptions();
-    return options.length > 0 ? options[options.length - 1] : null;
+  /** @internal */
+  get navItems(): Option[] {
+    return this.itemsStore.items;
   }
 
-  private getFirstSelectedOption(): Option | null {
-    return this.slottedListboxes[0]?.querySelector(`${OPTION_TAG_NAME}[selected]:not([disabled])`);
+  /** @internal */
+  private isValidItem(item: Element): boolean {
+    return item.matches(`${OPTION_TAG_NAME}:not([disabled])`);
   }
 
+  /** @internal */
+  private getFirstSelectedOption(): Option | undefined {
+    return this.navItems.find(el => el.hasAttribute('selected'));
+  }
+
+  /** @internal */
+  private getFirstOption(): Option {
+    return this.navItems[0];
+  }
+
+  /**
+   * Update the tabIndex of the list items when a new item is added.
+   *
+   * @internal
+   */
+  private handleCreatedEvent = (event: Event) => {
+    const createdElement = event.target as HTMLElement;
+    if (!this.isValidItem(createdElement)) {
+      return;
+    }
+
+    createdElement.tabIndex = -1;
+  };
+
+  /**
+   * Update the focus when an item is removed.
+   * If there is a next item, focus it. If not, focus the previous item.
+   *
+   * @internal
+   */
+  private handleDestroyEvent = (event: Event) => {
+    const destroyedElement = event.target as HTMLElement;
+    if (!this.isValidItem(destroyedElement) || destroyedElement.tabIndex !== 0) {
+      return;
+    }
+
+    const destroyedItemIndex = this.navItems.findIndex(node => node === destroyedElement);
```

Comment:
`findIndex` is for search based on generic predicate, but here you just compare references and `indexOf` is enough for that:

```suggestion
const destroyedItemIndex = this.navItems.indexOf(destroyedElement);
```